File: app/apps/file_editor_cm6/monaco_editor/editor_rpc_contract.py
# pyright: strict
from __future__ import annotations

from typing import Final, Literal, TypedDict, cast

from ..socketio_jsonrpc import (
    JsonRpcEnvelopeError,
    coerce_jsonrpc_envelope,
    is_jsonrpc_id,
    normalize_jsonrpc_params,
)

JSONRPC_VERSION: Final = "2.0"
EDITOR_RPC_EVENT: Final = "rpc"

JSONRPC_PARSE_ERROR: Final = -32700
JSONRPC_INVALID_REQUEST: Final = -32600
JSONRPC_METHOD_NOT_FOUND: Final = -32601
JSONRPC_INVALID_PARAMS: Final = -32602
JSONRPC_INTERNAL_ERROR: Final = -32603
JSONRPC_APPLICATION_ERROR: Final = -32000

JsonRpcId = str | int

EDITOR_RPC_METHOD_OPEN: Final = "editor.open"
EDITOR_RPC_METHOD_JUMP_TO_LINE: Final = "editor.jumpToLine"
EDITOR_RPC_METHOD_GIT_BASELINES_GET: Final = "editor.gitBaselines.get"
EDITOR_RPC_METHOD_DRAFT_DIFF_GET: Final = "editor.draftDiff.get"
EDITOR_RPC_METHOD_MIRROR_PUBLISH: Final = "editor.mirror.publish"
EDITOR_RPC_METHOD_SAVE: Final = "editor.save"
EDITOR_RPC_METHOD_MENTION_REQUEST: Final = "editor.mention.request"
EDITOR_RPC_METHOD_HOST_SAVE: Final = "editor.host.save"
EDITOR_RPC_METHOD_FOCUS: Final = "editor.focus"
EDITOR_RPC_METHOD_BLUR: Final = "editor.blur"
EDITOR_RPC_METHOD_READY_PUBLISH: Final = "editor.ready.publish"
EDITOR_RPC_METHOD_CACHE_STATE_PUBLISH: Final = "editor.cacheState.publish"
EDITOR_RPC_METHOD_DRAFT_STATE_PUBLISH: Final = "editor.draftState.publish"
EDITOR_RPC_METHOD_NOTIFY_PUBLISH: Final = "editor.notify.publish"
EDITOR_RPC_METHOD_OPEN_COMPLETE_PUBLISH: Final = "editor.openComplete.publish"
EDITOR_RPC_METHOD_DIAGNOSTICS_COUNTS_PUBLISH: Final = "editor.diagnosticsCounts.publish"
EDITOR_RPC_METHOD_SCROLL_STATE_PUBLISH: Final = "editor.scrollState.publish"
EDITOR_RPC_METHOD_MODEL_READY: Final = "editor.modelReady"
EDITOR_RPC_METHOD_SAVE_SNAPSHOT_RESPONSE: Final = "editor.save.snapshot.response"
EDITOR_RPC_METHOD_ISSUES_DUMP_RESPONSE: Final = "editor.issues.dump.response"
EDITOR_RPC_METHOD_BREADCRUMB_NAVIGATE: Final = "editor.breadcrumb.navigate"

EditorRpcMethod = Literal[
    "editor.open",
    "editor.jumpToLine",
    "editor.gitBaselines.get",
    "editor.draftDiff.get",
    "editor.mirror.publish",
    "editor.save",
    "editor.mention.request",
    "editor.host.save",
    "editor.focus",
    "editor.blur",
    "editor.ready.publish",
    "editor.cacheState.publish",
    "editor.draftState.publish",
    "editor.notify.publish",
    "editor.openComplete.publish",
    "editor.diagnosticsCounts.publish",
    "editor.scrollState.publish",
    "editor.modelReady",
    "editor.save.snapshot.response",
    "editor.issues.dump.response",
    "editor.breadcrumb.navigate",
]

EDITOR_RPC_NOTIFICATION_STATE_SSOT: Final = "editor.state.ssot"
EDITOR_RPC_NOTIFICATION_FILE_OPENED: Final = "editor.file.opened"
EDITOR_RPC_NOTIFICATION_FILE_JUMP_TO_LINE: Final = "editor.file.jumpToLine"
EDITOR_RPC_NOTIFICATION_MIRROR_UPDATED: Final = "editor.mirror.updated"
EDITOR_RPC_NOTIFICATION_GIT_BASELINES: Final = "editor.gitBaselines.updated"
EDITOR_RPC_NOTIFICATION_DRAFT_DIFF: Final = "editor.draftDiff.updated"
EDITOR_RPC_NOTIFICATION_PREFS_CHANGED: Final = "editor.prefs.changed"
EDITOR_RPC_NOTIFICATION_CACHE_STATE: Final = "editor.cache.state"
EDITOR_RPC_NOTIFICATION_DRAFT_STATE: Final = "editor.draft.state"
EDITOR_RPC_NOTIFICATION_READY: Final = "editor.ready"
EDITOR_RPC_NOTIFICATION_NOTIFY: Final = "editor.notify"
EDITOR_RPC_NOTIFICATION_OPEN_COMPLETE: Final = "editor.open.complete"
EDITOR_RPC_NOTIFICATION_DIAGNOSTICS: Final = "editor.diagnostics.updated"
EDITOR_RPC_NOTIFICATION_DIAGNOSTICS_COUNTS: Final = "editor.diagnostics.counts"
EDITOR_RPC_NOTIFICATION_ADAPTER_STATE: Final = "editor.adapter.state"
EDITOR_RPC_NOTIFICATION_SEMANTIC_TOKENS_PROVIDER_REGISTERED: Final = "editor.semanticTokens.providerRegistered"
EDITOR_RPC_NOTIFICATION_ISSUES_DUMP_REQUEST: Final = "editor.issues.dump.request"
EDITOR_RPC_NOTIFICATION_ISSUES_DUMP_RESPONSE: Final = "editor.issues.dump.response"
EDITOR_RPC_NOTIFICATION_SAVE_SNAPSHOT_REQUEST: Final = "editor.save.snapshot.request"
EDITOR_RPC_NOTIFICATION_ISSUES_COMMAND: Final = "editor.issues.command"
EDITOR_RPC_NOTIFICATION_FIND_COMMAND: Final = "editor.find.command"
EDITOR_RPC_NOTIFICATION_OPEN_STATE_CHANGED: Final = "editor.openState.changed"

EditorRpcNotification = Literal[
    "editor.state.ssot",
    "editor.file.opened",
    "editor.file.jumpToLine",
    "editor.mirror.updated",
    "editor.gitBaselines.updated",
    "editor.draftDiff.updated",
    "editor.prefs.changed",
    "editor.cache.state",
    "editor.draft.state",
    "editor.ready",
    "editor.notify",
    "editor.open.complete",
    "editor.diagnostics.updated",
    "editor.diagnostics.counts",
    "editor.adapter.state",
    "editor.semanticTokens.providerRegistered",
    "editor.issues.dump.request",
    "editor.issues.dump.response",
    "editor.save.snapshot.request",
    "editor.issues.command",
    "editor.find.command",
    "editor.openState.changed",
]


class JsonRpcErrorObject(TypedDict, total=False):
    code: int
    message: str
    data: dict[str, object]


class JsonRpcRequestEnvelope(TypedDict):
    jsonrpc: Literal["2.0"]
    id: JsonRpcId
    method: EditorRpcMethod
    params: dict[str, object]


class JsonRpcNotificationEnvelope(TypedDict):
    jsonrpc: Literal["2.0"]
    method: str
    params: dict[str, object]


class JsonRpcSuccessEnvelope(TypedDict):
    jsonrpc: Literal["2.0"]
    id: JsonRpcId
    result: object


class JsonRpcErrorEnvelope(TypedDict):
    jsonrpc: Literal["2.0"]
    id: JsonRpcId | None
    error: JsonRpcErrorObject


class EditorRpcProtocolError(Exception):
    def __init__(self, code: int, message: str, *, data: dict[str, object] | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.data = data or {}


class EditorRpcDispatchError(Exception):
    def __init__(self, code: int, message: str, *, data: dict[str, object] | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.data = data or {}


def coerce_jsonrpc_request_envelope(payload: object) -> JsonRpcRequestEnvelope | None:
    try:
        envelope = coerce_jsonrpc_envelope(payload)
    except JsonRpcEnvelopeError as exc:
        raise EditorRpcProtocolError(exc.code, exc.message) from exc
    if not envelope.has_id or not is_jsonrpc_id(envelope.request_id):
        return None
    params_obj: object = envelope.params
    if params_obj is not None and not isinstance(params_obj, dict):
        raise EditorRpcProtocolError(JSONRPC_INVALID_PARAMS, "params_must_be_object")
    return {
        "jsonrpc": JSONRPC_VERSION,
        "id": cast(JsonRpcId, envelope.request_id),
        "method": cast(EditorRpcMethod, envelope.method),
        "params": normalize_jsonrpc_params(cast(object, params_obj)),
    }


def coerce_jsonrpc_notification_envelope(payload: object) -> JsonRpcNotificationEnvelope:
    try:
        envelope = coerce_jsonrpc_envelope(payload)
    except JsonRpcEnvelopeError as exc:
        raise EditorRpcProtocolError(exc.code, exc.message) from exc
    params_obj: object = envelope.params
    if params_obj is not None and not isinstance(params_obj, dict):
        raise EditorRpcProtocolError(JSONRPC_INVALID_PARAMS, "params_must_be_object")
    return {
        "jsonrpc": JSONRPC_VERSION,
        "method": envelope.method,
        "params": normalize_jsonrpc_params(cast(object, params_obj)),
    }

File: app/apps/file_editor_cm6/monaco_editor/editor_rpc_contract.ts
export const JSONRPC_VERSION = '2.0' as const;
export const EDITOR_RPC_EVENT = 'rpc' as const;

export type JsonRpcId = string | number;

export interface JsonRpcErrorObject {
  code: number;
  message: string;
  data?: Record<string, unknown>;
}

export interface JsonRpcRequestEnvelope {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId;
  method: EditorRpcMethodName;
  params: Record<string, unknown>;
}

export interface JsonRpcNotificationEnvelope {
  jsonrpc: typeof JSONRPC_VERSION;
  method: EditorRpcMethodName | EditorRpcNotificationName;
  params: Record<string, unknown>;
}

export interface JsonRpcSuccessEnvelope {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId;
  result: unknown;
}

export interface JsonRpcErrorEnvelope {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId | null;
  error: JsonRpcErrorObject;
}

export const EDITOR_RPC_METHODS = {
  open: 'editor.open',
  jumpToLine: 'editor.jumpToLine',
  gitBaselinesGet: 'editor.gitBaselines.get',
  draftDiffGet: 'editor.draftDiff.get',
  mirrorPublish: 'editor.mirror.publish',
  save: 'editor.save',
  mentionRequest: 'editor.mention.request',
  hostSave: 'editor.host.save',
  focus: 'editor.focus',
  blur: 'editor.blur',
  readyPublish: 'editor.ready.publish',
  cacheStatePublish: 'editor.cacheState.publish',
  draftStatePublish: 'editor.draftState.publish',
  notifyPublish: 'editor.notify.publish',
  openCompletePublish: 'editor.openComplete.publish',
  diagnosticsCountsPublish: 'editor.diagnosticsCounts.publish',
  scrollStatePublish: 'editor.scrollState.publish',
  modelReady: 'editor.modelReady',
  saveSnapshotResponse: 'editor.save.snapshot.response',
  issuesDumpResponse: 'editor.issues.dump.response',
  breadcrumbNavigate: 'editor.breadcrumb.navigate',
} as const;

export type EditorRpcMethodName = (typeof EDITOR_RPC_METHODS)[keyof typeof EDITOR_RPC_METHODS];

export const EDITOR_RPC_NOTIFICATIONS = {
  stateSsot: 'editor.state.ssot',
  fileOpened: 'editor.file.opened',
  fileJumpToLine: 'editor.file.jumpToLine',
  mirrorUpdated: 'editor.mirror.updated',
  gitBaselines: 'editor.gitBaselines.updated',
  draftDiff: 'editor.draftDiff.updated',
  prefsChanged: 'editor.prefs.changed',
  cacheState: 'editor.cache.state',
  draftState: 'editor.draft.state',
  ready: 'editor.ready',
  notify: 'editor.notify',
  openComplete: 'editor.open.complete',
  diagnostics: 'editor.diagnostics.updated',
  diagnosticsCounts: 'editor.diagnostics.counts',
  adapterState: 'editor.adapter.state',
  semanticTokensProviderRegistered: 'editor.semanticTokens.providerRegistered',
  issuesDumpRequest: 'editor.issues.dump.request',
  issuesDumpResponse: 'editor.issues.dump.response',
  saveSnapshotRequest: 'editor.save.snapshot.request',
  issuesCommand: 'editor.issues.command',
  findCommand: 'editor.find.command',
  editCommand: 'editor.edit.command',
  openStateChanged: 'editor.openState.changed',
  projectSwitching: 'editor.project.switching',
  projectSwitched: 'editor.project.switched',
} as const;

export type EditorRpcNotificationName = (typeof EDITOR_RPC_NOTIFICATIONS)[keyof typeof EDITOR_RPC_NOTIFICATIONS];

export interface EditorRpcOpenParams {
  path: string;
  line?: number;
  column?: number;
  request_id?: string;
}

export interface EditorRpcJumpToLineParams {
  line: number;
  column?: number;
  focus?: boolean;
  scroll_y?: string;
  scroll_to_top?: boolean;
}

export interface EditorRpcGitBaselinesGetParams {
  path: string;
}

export interface EditorRpcDraftDiffGetParams {
  path: string;
  requestId?: string;
  request_id?: string;
}

export interface EditorRpcMirrorPublishParams {
  path: string;
  content: string;
  content_sha256?: string;
  base_sha256?: string;
}

export interface EditorRpcSaveParams {
  path: string;
  content?: string;
  content_sha256?: string;
  base_sha256?: string;
  request_id?: string;
  requestId?: string;
}

export interface EditorRpcMentionRequestParams {
  path: string;
  lineNo?: number;
  col?: number;
  endLineNo?: number;
  endCol?: number;
  content?: string;
}

export interface EditorRpcStateSsotNotificationParams {
  project?: string | null;
  session_state?: Record<string, unknown>;
  preferences?: Record<string, unknown>;
  currentPath?: string | null;
  file?: Record<string, unknown>;
}

export interface EditorRpcFileOpenedNotificationParams extends Record<string, unknown> {
  path?: string;
  request_id?: string;
}

export interface EditorRpcGitBaselinesNotificationParams extends Record<string, unknown> {
  path?: string;
}

export interface EditorRpcDraftDiffNotificationParams extends Record<string, unknown> {
  path?: string;
  hunks?: unknown[];
}

export interface EditorRpcPrefsChangedNotificationParams extends Record<string, unknown> {
  source_client?: string;
}

export function buildEditorRpcRequestEnvelope(
  id: JsonRpcId,
  method: EditorRpcMethodName,
  params: Record<string, unknown>,
): JsonRpcRequestEnvelope {
  return {
    jsonrpc: JSONRPC_VERSION,
    id,
    method,
    params,
  };
}

export function buildEditorRpcNotificationEnvelope(
  method: EditorRpcMethodName | EditorRpcNotificationName,
  params: Record<string, unknown>,
): JsonRpcNotificationEnvelope {
  return {
    jsonrpc: JSONRPC_VERSION,
    method,
    params,
  };
}

export function isJsonRpcSuccessEnvelope(value: unknown): value is JsonRpcSuccessEnvelope {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  return record.jsonrpc === JSONRPC_VERSION && 'id' in record && 'result' in record;
}

export function isJsonRpcErrorEnvelope(value: unknown): value is JsonRpcErrorEnvelope {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  return record.jsonrpc === JSONRPC_VERSION && 'error' in record;
}

export function isJsonRpcNotificationEnvelope(value: unknown): value is JsonRpcNotificationEnvelope {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  return record.jsonrpc === JSONRPC_VERSION && typeof record.method === 'string' && !('id' in record);
}
